lms/services/digest.py

```diff
@@ -0,0 +1,264 @@
+        self, h_userids, since, until, override_to_email=None
+    ):
+        annotations = self.h_api.get_annotations(h_userids, since, until)
+        unified_courses = self.helper.unified_courses(annotations)
+
+        for h_userid in h_userids:
+            unified_user = UnifiedUser.make(self.db, h_userid)
+            digest = self.helper.instructor_digest(unified_user, unified_courses)
+
+            if not digest["num_annotations"]:
+                continue
+
+            if override_to_email is None:
+                to_email = unified_user.email
+            else:
+                to_email = override_to_email
+
+            self.mailchimp_service.send_template(
+                "instructor-email-digest",
+                self.sender,
+                recipient=EmailRecipient(to_email, unified_user.display_name),
+                template_vars=digest,
+            )
+
+
+@dataclass
+class UnifiedUser:
+    """All User's for a given h_userid, unified across all ApplicationInstance's."""
+
+    users: List[User] = field(default_factory=list)
+
+    def __post_init__(self):
+        for user in self.users:
+            assert user.h_userid == self.h_userid
+
+    @property
+    def h_userid(self) -> str:
+        """Return this user's h_userid."""
+        return self.users[0].h_userid
+
+    @property
+    def user_ids(self) -> List[int]:
+        """Return the User.id's of all the users with self.h_userid."""
+        return [user.id for user in self.users]
+
+    @property
+    def email(self) -> Optional[str]:
+        """Return an email address for this user."""
+        for user in self.users:
+            if user.email:
+                return user.email
+
+        return None
+
+    @property
+    def display_name(self) -> Optional[str]:
+        """Return a display name for this user."""
+        for user in self.users:
+            if user.display_name:
+                return user.display_name
+
+        return None
+
+    @classmethod
+    def make(cls, db, h_userid):
+        """Return a UnifiedUser for the given h_userid."""
+        return cls(
+            db.scalars(
+                select(User).filter_by(h_userid=h_userid).order_by(User.updated.desc())
+            ).all()
+        )
+
+
+@dataclass
+class UnifiedCourse:
+    """All Course's for a given authority_provided_id, unified across all ApplicationInstance's."""
+
+    courses: List[Course] = field(default_factory=list)
+    annotations: List[dict] = field(default_factory=list)
+
+    def __post_init__(self):
+        for course in self.courses:
+            assert course.authority_provided_id == self.authority_provided_id
+
+    @property
+    def authority_provided_id(self) -> str:
+        """Return this course's authority_provided_id."""
+        return self.courses[0].authority_provided_id
+
+    @property
+    def title(self) -> str:
+        """Return a title for this course."""
+        return self.courses[0].lms_name
```

Comment:
`select lms_name from grouping where lms_name is null or lms_name ='' and type = 'course'`

doesn't return any rows :+1:

Reply:
Yeah `lms_name` is not nullable (unlike user display names and emails). I guess I could watch our for empty strings but if there's non actually in the DB it seems unnecessary